src/ar_config.erl

```diff
@@ -0,0 +1,171 @@
+-module(ar_config).
+-export([parse/1]).
+-include("ar.hrl").
+-include("ar_config.hrl").
+
+%%%===================================================================
+%%% Public interface.
+%%%===================================================================
+
+parse(Config) when is_binary(Config) ->
+	case ar_serialize:json_decode(Config) of
+		{ok, JsonValue} -> parse_options(JsonValue);
+		{error, _} -> {error, bad_json, Config}
+	end.
+
+%%%===================================================================
+%%% Private functions.
+%%%===================================================================
+
+
+parse_options({KVPairs}) when is_list(KVPairs) ->
+	parse_options(KVPairs, #opts{});
+parse_options(JsonValue) ->
+	{error, root_not_object, JsonValue}.
+
+parse_options([{_, null} | Rest], Opts) ->
+	parse_options(Rest, Opts);
+parse_options([{<<"config_file">>, _} | _], _) ->
+	{error, config_file_set};
+parse_options([{<<"peers">>, Peers} | Rest], Opts) when is_list(Peers) ->
+	ParsedPeers = lists:map(fun(Peer) -> ar_util:parse_peer(Peer) end, Peers),
+	parse_options(Rest, Opts#opts { peers = ParsedPeers });
+parse_options([{<<"peers">>, Peers} | _], _) ->
+	{error, {bad_type, peers, array}, Peers};
+parse_options([{<<"start_hash_list">>, Hash} | Rest], Opts) when is_binary(Hash) ->
+	parse_options(Rest, Opts#opts { start_hash_list = ar_util:decode(Hash) });
+parse_options([{<<"start_hash_list">>, Hash} | _], _) ->
+	{error, {bad_type, start_hash_list, string}, Hash};
```

Comment:
What's the point checking for some errors when the `ar_util:decode/1` will crash if the value is not in a valid format? Also, being this specific explaining that it's `bad_type` and something about `string` (which btw is not true - it can be anything but a binary) is not really needed.

I'm suggesting this:
Create a function that parses one argument (e.g. "start_hash_list" and the hash). It can return an error tuple for specific error handling. If it crashes, then `parse/1` can return something like `{error, {bad_config, "start_hash_list"}}`

Reply:
So the types I'm referring to here (such as `string`) are the expected JSON types. My thinking was again that the caller of this function could turn the error tuple into a readable error message such as "start_hash_list should be a string".